### PR TITLE
[4.1] Cast client_id in template to integer to support mysql PDO

### PR DESCRIPTION
**Pull request by bembelimen**
Pull Request for Issue #37487 .

### Summary of Changes
See #37489 for the details

### CODE REVIEW

File: administrator/components/com_templates/src/Model/TemplateModel.php [hunk 1]
```diff
@@ -681,6 +681,9 @@ public function &getTemplate()
 			{
 				$this->template = $result;
 
+				// Client ID is on PDO mySQL not an iteger, so enforce here
```
Comment (richard67) [inline administrator/components/com_templates/src/Model/TemplateModel.php]:
```suggestion
				// Client ID is on PDO mySQL not an integer, so enforce here
```

**Comment by laoneo** [comment]
Is this also needed on the templateS model?

**Comment by bembelimen** [comment]
> Is this also needed on the templateS model?

The usage I found was once in the model itself where it was used only as key and in the default.php where it was already casted. So I would say "no", but in general we have this problem on many places I guess (with other fields)

**Comment by brianteeman** [comment]
> // Client ID is on PDO mySQL not an integer, so enforce here

This doesnt make sense. Not sure exactly what you are trying to say. Guessing at

// Client ID is not an integer, so enforce here

**Comment by brianteeman** [comment]
> in general we have this problem on many places I guess (with other fields)

Shouldnt they be addressed with this PR - at least the ones relating to the template

**Comment by bembelimen** [comment]
> > in general we have this problem on many places I guess (with other fields)
> 
> Shouldnt they be addressed with this PR - at least the ones relating to the template

Do you see any spontanious? It was more a general remark. But I'm happy to fix it, if there are some very obvious ones.

**Comment by alikon** [comment]
this is a funny one...
i'm not able to replicate it...just tested on a normal cfg  :smile: 
even if i'm on mysql pdo 
anyway 
1. the executed query is 
 ```sql
SELECT `extension_id`,`client_id`,`element`,`name`,`manifest_cache`
FROM `#__extensions`
WHERE `extension_id` = :pk AND `type` = 'template'
 ```

2. the client_id field is declared as _tinyint_ so cannot be **string** 
https://github.com/joomla/joomla-cms/blob/126ed6a1d4d4661d839096ca10bb2102486ae0dd/installation/sql/mysql/base.sql#L126

oh wait ...probably i'm not on MariaDB  ? but 
debug tell me this
debug

so it should be digged a bit more deep

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 8892cea242bd555f5a8f88cabd8b39526288fc94<br /><br />Tested successfully in Joomla 4.1.3-dev of 21 April in Wampserver 3.2.7 using MySQL 8.0.27 and PHP 8.0.15.

This PR fixes the same issue with MySQL (PDO) selection breaking the path to the template files as the already closed #37489 did - https://github.com/joomla/joomla-cms/pull/37489.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37497">issues.joomla.org/tracker/joomla-cms/37497</a>.</sub>

**Comment by Sandra97** [comment]
I have tested this item :white_check_mark: successfully on 8892cea242bd555f5a8f88cabd8b39526288fc94<br /><br />Tested on the site on which I faced the issue (4.1.2). Path is now correct. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37497">issues.joomla.org/tracker/joomla-cms/37497</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37497">issues.joomla.org/tracker/joomla-cms/37497</a>.</sub>

**Comment by alikon** [comment]
still a funny one, my observations... are simply still ignored
i give up

**Comment by richard67** [comment]
@alikon Well I was able to reproduce the issue. And in past I have observed that depending on the driver and the database kind and Version columns with integer data types had been converted to strings.

**Comment by alikon** [comment]
i doubt still, anyway....
probably mine are not valid...fine
go forward

**Comment by alikon** [comment]
mine reply was not in synch with your last one ....probably

**Comment by alikon** [comment]
still stand anyway... simply ignore my comments mine are lastly ... killed from someone's thumbs down --- even if I declare as  a DIRTY fix
Note to myself: stop doing things
:unicorn:

**Comment by richard67** [comment]
?

**Comment by alikon** [comment]
Can we have a chat in 15 min?


Il gio 21 apr 2022, 21:36 Richard Fath ***@***.***> ha
scritto:

**Comment by richard67** [comment]
On Glip? If necessary, ok

**Comment by alikon** [comment]
thanks @richard67  to listen

**Comment by richard67** [comment]
We should maybe dig deeper into what causes different behavior of different versions of MySQL and MariaDB, but for the mean time this PR as a quick and maybe dirty fix is ok, I think.

**Comment by alikon** [comment]
you know i'm the man for a quick & dirty fix
...in some tables --- are called workaround---

**Review by alikon — COMMENTED** [review]
should  be digger more deep

**Comment by laoneo** [comment]
So what is here the exact issue that it is considered a dirty hack?

**Comment by richard67** [comment]
> So what is here the exact issue that it is considered a dirty hack?

@laoneo Maybe the wrong words, but as long as we don't know why we get an integer or a string for an integer data type column depending on database type and version, it leaves me with the feeling that we fix the symptom and not the root cause. But that's just a gut feeling and I may be wrong.

**Comment by laoneo** [comment]
Valid concern.

**Comment by HLeithner** [comment]
The problem with the MySQL drivers are that it depends if PDO, MySQLi (MySQL) if it typecasts correctly and if we use prepared statements or not. it has been fixed (in a b/c incompatible way) in php 8.1 https://www.php.net/manual/en/migration81.incompatible.php#migration81.incompatible.pdo.mysql 

if we fix this here we should add a deprecation notice (or removal notice with J5, where the minimum version will be php 8.1). I'm surprise that this is the only problem found yet for this behavior.

**Comment by richard67** [comment]
I think @HLeithner 's [comment above](https://github.com/joomla/joomla-cms/pull/37497#issuecomment-1109482093) and the [link it refers to](https://www.php.net/manual/en/migration81.incompatible.php#migration81.incompatible.pdo.mysql) explain what happens.

For the mean time, until we have fixed the root cause, which we cleanly could do with J5 as mentioned by Harald, we will need such fixes like provided here.

Therefore I will merge this PR now.

**Comment by richard67** [comment]
Thanks.

**Pull request merged**